src/isc-dhcp/patch/0009-Support-for-dual-tor-scenario.patch

```diff
@@ -0,0 +1,437 @@
+From 4378beb41efa256eb2b492c0046ad071dbf17076 Mon Sep 17 00:00:00 2001
+From: Tianrong Zhang <trzhang@microsoft.com>
+Date: Wed, 25 Nov 2020 15:40:54 -0800
+Subject: [PATCH] Support-for-dual-tor-scenario
+
+---
+ relay/Makefile.am |   1 +
+ relay/dhcrelay.c  | 248 ++++++++++++++++++++++++++++++++++++++++------
+ 2 files changed, 217 insertions(+), 32 deletions(-)
+
+diff --git a/relay/Makefile.am b/relay/Makefile.am
+index 2ba5979..bb1c21d 100644
+--- a/relay/Makefile.am
++++ b/relay/Makefile.am
+@@ -1,4 +1,5 @@
+ AM_CPPFLAGS = -DLOCALSTATEDIR='"@localstatedir@"'
++AM_LDFLAGS = -pthread
+ 
+ sbin_PROGRAMS = dhcrelay
+ dhcrelay_SOURCES = dhcrelay.c
+diff --git a/relay/dhcrelay.c b/relay/dhcrelay.c
+index e158efe..f410a86 100644
+--- a/relay/dhcrelay.c
++++ b/relay/dhcrelay.c
+@@ -32,6 +32,7 @@
+ #include <sys/time.h>
+ #include <sys/ioctl.h>
+ #include <isc/file.h>
++#include <pthread.h>
+ 
+ TIME default_lease_time = 43200; /* 12 hours... */
+ TIME max_lease_time = 86400; /* 24 hours... */
+@@ -56,6 +57,7 @@ int bogus_agent_drops = 0;	/* Packets dropped because agent option
+ 				   specified. */
+ int bogus_giaddr_drops = 0;	/* Packets sent to us to relay back to a
+ 				   client, but with a bogus giaddr. */
++int bogus_yiaddr_drops = 0;
+ int client_packets_relayed = 0;	/* Packets relayed from client to server. */
+ int server_packet_errors = 0;	/* Errors sending packets to servers. */
+ int server_packets_relayed = 0;	/* Packets relayed from server to client. */
+@@ -83,6 +85,16 @@ int max_hop_count = 10;		/* Maximum hop count */
+ int no_daemon = 0;
+ int dfd[2] = { -1, -1 };
+ 
++int enable_support_for_dual_tor = 0;
++struct in_addr loopback_addr;
++int enable_receive_pkt_from_loopback = 0;
++int sockfd_loopback;
++
++struct downstream_intf_list {
++	struct downstream_intf_list *next;
++	struct interface_info *interface;
++} *downstream_intfs = NULL;
++
+ #ifdef DHCPv6
+ 	/* Force use of DHCPv6 interface-id option. */
+ isc_boolean_t use_if_id = ISC_FALSE;
+@@ -156,6 +168,9 @@ static int load_interface_alias_map(const char *port_alias_map_file_path);
+ static int get_interface_alias_by_name(const char *if_name, char *if_alias_out);
+ static void free_interface_alias_map(void);
+ 
++static void *receive_pkt_from_loopback(void *);
++static void free_downstream_intfs(void);
++
+ static const char copyright[] =
+ "Copyright 2004-2018 Internet Systems Consortium.";
+ static const char arr[] = "All rights reserved.";
+@@ -189,6 +204,7 @@ char *progname;
+ "                     [-iu interface0 [ ... -iu interfaceN]\n" \
+ "                     [-id interface0 [ ... -id interfaceN]\n" \
+ "                     [-U interface]\n" \
```

Comment:
I came to learn about -U option recently. Can we instead of having new option, provide the -U with the loopback interface? this will do the following:
```
		/* RFC3527: Use the inbound packet's interface address in
		 * the link selection suboption and set the outbound giaddr
		 * to the uplink address. */
		if (adding_link_select) {
			*sp++ = RAI_LINK_SELECT;
			*sp++ = 4u;
			memcpy(sp, &giaddr.s_addr, 4);
			sp += 4;
			packet->giaddr = uplink->addresses[0];
			log_debug ("Adding link selection suboption"
				   " with addr: %s", inet_ntoa(giaddr));
		}
```
which sending the `giaddr` as option 82 sub-option 5 and update `giaddr` with the loopback interface address.

Reply:
no, 1) loopback on sonic is a 'virtual' interface 2) we use different packet flow logic, e.g., we use yiaddr to filter vlan 3) two options for different scenarios, better not mix them up.

Reply:
This is not to say we are going to eliminate the filter logic. We would still be filtering the packets based on the `yiaddr` since the the loopback interface will be shared among different relay instances. Can you please elaborate on 1 and 3?

Reply:
dug a bit deeper in the code and it looks like -U would mark the interface as UPSTREAM and DOWNSTREAM (INTERFACE_STREAMS)
```
			interface_snorf(uplink, (INTERFACE_REQUESTED |
						INTERFACE_STREAMS));

```

We could use this information know where the packed is coming from, or we could following similar approach as `-U` option and mark the interface as UPSTREAM only. This way, there is no need to introduces a thread to receive packet. We will still be using the common lib to receive packets. let's sync up offline...

Reply:
> We could use this information know where the packed is coming from, or we could following similar approach as -U option and mark the interface as UPSTREAM only. This way, there is no need to introduces a thread to receive packet. We will still be using the common lib to receive packets.

I would prefer eliminating the need to spawn a new thread, if possible.